Initial Mistral implementation

OriPekelman opened the pull request:
Tried to follow existing implementation as much as possible and hopefully integrated correctly.

crmne commented:
Thank you @OriPekelman for your contribution!

Two new features to consider in your provider implementation:

1. **Model Aliases**: Please add entries for your provider in `aliases.json`:
   ```json
   "claude-3-5-sonnet": {
     "anthropic": "claude-3-5-sonnet-20241022",
     "your-provider": "your-provider-specific-id"
   }
   ```

2. **Provider Selection**: Users will be able to specify your provider:
   ```ruby
   chat = RubyLLM.chat(model: 'claude-3-5-sonnet', provider: 'your-provider')
   ```

Docs: https://rubyllm.com/guides/models#using-model-aliases

Please ask me for a code review when you're ready.

crmne commented:
@OriPekelman is this still on your radar? I'd love to merge Mistral support soon. Whenever you are ready, could you resolve the conflicts and rerequest a review? Thanks!

OriPekelman commented:
Yup, I'll try to get to it very soon.

crmne commented:
Hi @OriPekelman is Mistral OpenAI compatible?

OriPekelman commented:
> Hi @OriPekelman is Mistral OpenAI compatible?

Well, like most providers, it has some basic OpenAI compatability ... but only for chat completions - and even there you are going to have quite a bit of differences in tool calling and multi-turn behaviour. 

I rebased the whole thing and cleaned it up quite a bit + added  the `configure`. 

There are a number of tests I had to skip (for example Mistral does not support custom dimensions for embedding, the tool calling in multi-turn conversations doesn't support adding tools mid conversation, only the last chunk in a stream would have token count).

Hopefully this is useful.

crmne reviewed (changes requested):
Good stuff! I left you some comments

OriPekelman commented on lib/ruby_llm/providers/mistral/chat.rb:
```diff
@@ -0,0 +1,193 @@
+      module Chat
+        module_function
+
+        def completion_url
+          "#{Mistral.api_base(RubyLLM.config)}/chat/completions"
+        end
```
Addressed

OriPekelman commented on lib/ruby_llm/providers/mistral/chat.rb:
```diff
@@ -0,0 +1,193 @@
+              Array(tools)
+            end
+
+          puts "\n[DEBUG] Available tools: #{tools_array&.map { |t| t.name.to_s }}" if ENV["DEBUG"]
```
Addressed

OriPekelman commented on lib/ruby_llm/providers/mistral/chat.rb:
```diff
@@ -0,0 +1,193 @@
+              "none"
+            end
+
+          puts "[DEBUG] Tool choice: #{effective_tool_choice.inspect}" if ENV["DEBUG"]
```
Addressed

OriPekelman commented on spec/ruby_llm/chat_tools_spec.rb:
```diff
@@ -36,6 +36,7 @@ def execute
       model = model_info[:model]
       provider = model_info[:provider]
       it "#{provider}/#{model} can use tools" do # rubocop:disable RSpec/MultipleExpectations
+        skip 'Mistral does not reliably support tool usage' if provider == :mistral
```
OK I addressed the mistral too hallucination issue - from what I can see it only happens with the small model. At any rate it no longer breaks the implementation. It tends to believe it has a "web_search" tool whether you have one or not.

OriPekelman commented on spec/ruby_llm/embeddings_spec.rb:
```diff
@@ -23,6 +27,7 @@
       end
 
       it "#{provider}/#{model} can handle a single text with custom dimensions" do # rubocop:disable RSpec/MultipleExpectations
+        skip "Mistral embed does not support custom dimensions" if model == "mistral-embed"
```
Went for provider, but if I am not mistaken it is a string here.

OriPekelman commented:
Oh and now i'll try to do a clean rebase.

OriPekelman commented:
I see there are some new tests, I'll see if I can implement the local image stuff.

OriPekelman commented:
OK this should work now (given that models.json is up-to-date) for the failing vision tests.

OriPekelman commented:
OK rebased on main I did not commit the changes to models.json (as it is a source of possible merge conflicts) - but the update  seems to work .. and the tests look green after I run `rake models:update` I ran the test with both mistral small and medium.

crmne commented on lib/ruby_llm/providers/mistral.rb:
I'm confused as to why we need so much code for the Mistral provider. Isn't it OpenAI compatible? In that case, please take a look at already existing OpenAI compatible providers like Ollama and OpenRouter.

crmne commented on lib/tasks/capability_scraper.rb:
I guess this is a merge error, but we don't need this anymore thanks to Parsera.

crmne commented on lib/tasks/model_updater.rb:
This looks like a merge error too.

OriPekelman commented on lib/ruby_llm.rb:
```diff
@@ -18,7 +18,7 @@
   'deepseek' => 'DeepSeek',
   'bedrock' => 'Bedrock',
   'openrouter' => 'OpenRouter',
-  'pdf' => 'PDF'
```
OOps. That was part of me removing the PDF support for mistral (I added it with their dedicated end-point https://docs.mistral.ai/capabilities/OCR/basic_ocr/ but decided against it before the main functionality gets merged. My excuses.

OriPekelman commented on lib/ruby_llm.rb:
```diff
@@ -18,7 +18,7 @@
   'bedrock' => 'Bedrock',
   'openrouter' => 'OpenRouter',
-  'pdf' => 'PDF'
+  'mistral' => 'Mistral'
```
OK will remove that one too.

crmne commented:
Went ahead and implemented it in 92d5451b797879b1213556a5e53533ba69af6e58
